libportal/notification.c

```diff
@@ -60,22 +70,155 @@ ensure_action_invoked_connection (XdpPortal *portal)
 }
 
 typedef struct {
-  XdpPortal *portal;
-  GAsyncReadyCallback callback;
-  gpointer data;
-} CallDoneData;
+  GUnixFDList *fd_list;
+  GVariantIter *iter;
+  GVariantBuilder *builder;
+} CallData;
+
+static CallData*
+call_data_new (GVariantIter    *iter,
+               GVariantBuilder *builder)
+{
+  CallData *data;
+
+  data = g_new0 (CallData, 1);
+  data->fd_list = g_unix_fd_list_new ();
+  data->iter = iter;
+  data->builder = builder;
+
+  return data;
+}
+
+static void
+call_data_free (gpointer user_data)
+{
+  CallData *data = user_data;
+
+  g_clear_object (&data->fd_list);
+  g_clear_pointer (&data->iter, g_variant_iter_free);
+  g_clear_pointer (&data->builder, g_variant_builder_unref);
+
+  g_free (data);
+}
+
+static void
+call_done_cb (GObject      *source,
+              GAsyncResult *result,
+              gpointer      data)
+{
+  g_autoptr(GTask) task = G_TASK (data);
+  XdpPortal *portal = g_task_get_source_object (task);
+  g_autoptr(GError) error = NULL;
+  g_autoptr(GVariant) res = NULL;
+
+  res = g_dbus_connection_call_with_unix_fd_list_finish (portal->bus, NULL, result, &error);
+
+  if (res)
+    g_task_return_boolean (task, TRUE);
+  else
+    g_task_return_error (task, g_steal_pointer (&error));
+}
+
+static void
+parse_notification (GTask *task)
```

Comment:
It's a bit unexpected that a function named `parse_notification` does a dbus call. Maybe split the actual "parsing" out of here.

Reply:
i can rename the function to `handle_notification()`